### PR TITLE
Add S3-specific Prometheus metrics

### PR DESCRIPTION
**Pull request by quodlibetor**
Per-bucket metrics on objects discovered and downloaded, and bytes/lines read.

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/5740)

### CODE REVIEW

File: src/dataflow/src/source/s3/metrics.rs [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+            )
+            .unwrap();
+            static ref BYTES_DOWNLOADED: UIntCounterVec = register_uint_counter_vec!(
+                "mz_s3_bytes_downloaded",
+                "The total count of bytes downloaded for this source",
+                LABELS
+            )
+            .unwrap();
+            static ref LINES_INGESTED: UIntCounterVec = register_uint_counter_vec!(
+                "mz_s3_lines_ingested",
```
Comment (cirego) [inline src/dataflow/src/source/s3/metrics.rs:37]: Is it lines or messages? Can messages be more than one line long?

Comment (quodlibetor) [reply src/dataflow/src/source/s3/metrics.rs:37]: it is lines, currently all text sources are line-delimited. We treat lines as messages, though, so perhaps messages is the more future-proof term.

Comment (cirego) [reply src/dataflow/src/source/s3/metrics.rs:37]: Sounds good -- thanks for the explanation. I was confused because both `lines` and `messages` were used.

**Comment by quodlibetor** [comment]
I'm adding another couple metrics, so switching this to draft.

**Comment by quodlibetor** [comment]
ready for review again.

File: src/dataflow/src/source/s3/metrics.rs [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use lazy_static::lazy_static;
+use prometheus::core::AtomicU64;
+use prometheus::{register_uint_counter_vec, DeleteOnDropCounter, UIntCounterVec};
+
+pub(super) struct BucketMetrics {
+    objects_downloaded: DeleteOnDropCounter<'static, AtomicU64>,
+    pub objects_duplicate: DeleteOnDropCounter<'static, AtomicU64>,
```
Comment (cirego) [inline src/dataflow/src/source/s3/metrics.rs:16]: Why is this one `pub`? Is this intended to be used outside of this file? I don't see it mentioned elsewhere in the review and `inc` is not incrementing this counter.

**Review by cirego — APPROVED** [review]

File: src/dataflow/src/source/s3/metrics.rs (hunk 2 above)
Comment (quodlibetor) [reply src/dataflow/src/source/s3/metrics.rs:17]: it is used in `source/s3.rs` on line 206, I've added a comment explaining why it's different.

**Pull request merged**